### PR TITLE
Add HW3 tests by thomas0829

### PR DESCRIPTION
**Pull request by thomas0829**
- Added 4 comprehensive tests (smoke, one-shot, edge, pattern)
- All tests passing (4/4)

### CODE REVIEW

File: hw3-test/test_hw3_thomas0829.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    author: thomas0829
+    reviewer: Chase-Zou
+    category: one-shot test
+    
+    Test normalization with specific known input/output pair
+    """
+    # Testing that 75 normalized by max 150 equals exactly 0.5
+    result = normalize(75, 150)
+    assert result == 0.5, f"Expected 0.5 but got {result}"
```
Comment (Chase-Zou) [inline hw3-test/test_hw3_thomas0829.py]: Nice one-shot test. Clear expected value and precise check.
Optional: consider adding a boundary test (e.g., value = 0 or value > max_value) in the future.

File: hw3-test/test_hw3_thomas0829.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    Test edge case: project created at the exact current moment
+    """
+    # Create a timestamp for right now
+    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+    
+    # Age factor for brand new project should be very low
+    score = age_factor(now)
+    
+    # Brand new projects (0 days old) should score very low
+    assert 0 <= score <= 0.3, f"Brand new project should score <= 0.3, got {score}"
```
Comment (Chase-Zou) [inline hw3-test/test_hw3_thomas0829.py:65]: Checking the “new project” case is meaningful.
One small note: tests depending on datetime.now() may introduce slight timing variance, though for HW3 this is acceptable.

File: hw3-test/test_hw3_thomas0829.py (hunk 1 above)
Comment (Chase-Zou) [inline hw3-test/test_hw3_thomas0829.py:47]: Checking the “new project” case is meaningful.
One small note: tests depending on datetime.now() may introduce slight timing variance, though for HW3 this is acceptable.

File: hw3-test/test_hw3_thomas0829.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    author: thomas0829
+    reviewer: Wenshu0206
+    category: smoke test
+    
+    Basic smoke test to ensure TokenManager initializes without crashing
+    """
+    tokens = ['ghp_test_token_1', 'ghp_test_token_2']
+    tm = TokenManager(tokens)
+    assert tm is not None
+    assert tm.get_token_count() == 2
```
Comment (Wenshu0206) [inline hw3-test/test_hw3_thomas0829.py:34]: This is a clean and straightforward smoke test. It correctly verifies that TokenManager can be instantiated without errors and that the initial token count matches the input. The test is easy to read and serves its purpose well.

A couple of optional improvements:
- Using len(tokens) in the assertion would make the test slightly more robust.
- Adding a @pytest.mark.smoke marker could better categorize the test within pytest.

File: hw3-test/test_hw3_thomas0829.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        project = base_project.copy()
+        project['stars'] = stars
+        score = calculate_github_score(project, 20000, 5000, 1000)
+        scores.append(score)
+    
+    # Verify pattern: scores should be non-decreasing
+    for i in range(len(scores) - 1):
+        assert scores[i] <= scores[i + 1], \
+            f"Score should increase with stars: {scores[i]} (stars={star_counts[i]}) > {scores[i + 1]} (stars={star_counts[i + 1]})"
+
```
Comment (Wenshu0206) [inline hw3-test/test_hw3_thomas0829.py:100]: This test is well-structured and clearly expresses its purpose: verifying that the GitHub score increases as star counts rise while keeping other project attributes constant. The use of a base project template avoids duplication, and the increasing star_counts list makes the pattern test easy to follow. The assertion message is descriptive and helpful for debugging.

**Review by Wenshu0206 — APPROVED** [review]
